File: C#/src/Implementations/AutoSaveManager.cs
using System;
using System.Threading.Tasks;

public class AutoSaveManager : ISaveAsyncManager
{
    private Task saveTask;
    private bool isSaving;
    private int saveCounter;
    private SaveJSONManager saver;
    private GameState gameState;

    public AutoSaveManager(SaveJSONManager saverRef, GameState gameState)
    {
        saver = saverRef;
        this.gameState = gameState;
        isSaving = false;
        saveCounter = 0;
    }

    public void SetSaveCounter(int newSaveCounter)
    {
        saveCounter = newSaveCounter;
    }

    public string SaveGameStateAsync(GameState gameState, string autoSaveFileName)
    {
        isSaving = true;
        SetSaveCounter(GetSaveCounter() + 1);

        string finalFileName = string.IsNullOrEmpty(autoSaveFileName)
            ? "autosave" + GetSaveCounter().ToString() + ".json"
            : autoSaveFileName;

        string fileNameForTask = finalFileName;
        var localGameState = gameState;
        var localSaver = saver;

        saveTask = Task.Run(async () =>
        {
            await Task.Delay(500);

            localSaver.SaveGameState(localGameState, fileNameForTask);

            isSaving = false;
            Console.WriteLine($"Autosave wykonany: {fileNameForTask}");
        });

        return finalFileName;
    }

    public int GetSaveCounter()
    {
        return saveCounter;
    }

    public bool IsSaving()
    {
        return isSaving;
    }
}

File: C++/src/Implementations/AutoSaveManager.cpp
#include "AutoSaveManager.h"
#include "SaveJSONManager.h"
#include "GameState.h"
#include <string>
#include <chrono>
#include <ctime>
#include <future>
#include <thread>
#include <cstdint>

AutoSaveManager::AutoSaveManager(ISaveSyncManager& saverRef) 
    : saver(saverRef) {}

void AutoSaveManager::SaveGameStateAsync(const GameState& state)
{
    isSaving.store(true);
	SetSaveCounter(GetSaveCounter() + 1);
    std::string autoSaveFileName = "autosave" + std::to_string(GetSaveCounter());
    saveFuture = std::async(std::launch::async, [this, &state, autoSaveFileName]()
    {
        GameState snapshot;
        snapshot.CopyStateToSnapshot(state);
        saver.SaveGameState(snapshot, autoSaveFileName);
        isSaving.store(false);
    });
}

bool AutoSaveManager::IsSaving() const { return isSaving; }

void AutoSaveManager::SetSaveCounter(uint8_t newSaveCounter) { saveCounter = newSaveCounter; }

uint8_t AutoSaveManager::GetSaveCounter() const { return saveCounter; }